src/support/time.c

```diff
@@ -67,3 +70,99 @@ __wt_seconds(WT_SESSION_IMPL *session, time_t *timep)
 
 	*timep = t.tv_sec;
 }
+
+/*
+ * __clock_server --
+ *	Within WiredTiger there are a number of places where we wish to time
+ *	the execution of operations, but the cost of calling time() is too high.
+ *	To work around this issue, we create a clock server below which keeps
+ *	time on the connection accessible globally.
+ *
+ * 	The clock server aims to keep time to roughly the precision of the
+ * 	TIMER_PRECISION_MS value and we accept that sometimes things may skew
+ *	until the next reset.
+ */
+static WT_THREAD_RET
+__clock_server(void *arg)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_SESSION_IMPL *session;
+	uint64_t timer_runs;
+
+	session = arg;
+	conn = S2C(session);
+	timer_runs = 0;
+
+	while (F_ISSET(conn, WT_CONN_SERVER_CLOCK)) {
+		/*
+		 * We re-capture timing the long way every so often to ensure
+		 * that we don't skew too much in our timing.
+		 */
+		if (timer_runs % TIMER_RECHECK_PERIOD == 0) {
+			__wt_epoch(NULL, &conn->epoch_clock);
+		} else {
+			/*
+			 * Fuzzy time tracking means we don't mind loosing some
+			 * precision when incrementing the time here.
+			 */
+			if (conn->epoch_clock.tv_nsec >=
+			    1000000000 - (TIMER_PRECISION_MS * WT_THOUSAND)) {
+				conn->epoch_clock.tv_sec++;
+				conn->epoch_clock.tv_nsec = 0;
+			} else {
+				conn->epoch_clock.tv_nsec
+				    += TIMER_PRECISION_MS * WT_THOUSAND;
+			}
+		}
+
+		__wt_sleep(0, TIMER_PRECISION_MS);
```

Comment:
Wouldn't we context switch and that itself in most cases be several micro secs. Lets assume we want 1 microsec precision, and we reset every 25 calls. If a single context switch takes say 2 microsecs (totally making this value up after googling context switch times), we can be off by as much as 50 microsecs, not too good if we are aiming for a 1 microsec precision. Need to think more about it. Initially I thought minimum precision is a millisec and context switch can likely be ignored. Not with microsecs I suppose.